source/common/event/scaled_range_timer.h

```diff
@@ -0,0 +1,69 @@
+#include <chrono>
+
+#include "envoy/event/timer.h"
+
+#include "absl/container/flat_hash_map.h"
+
+namespace Envoy {
+namespace Event {
+
+class ScaledRangeTimerManager {
+public:
+  ScaledRangeTimerManager(
+      Dispatcher& dispatcher, float scale_factor,
+      std::chrono::milliseconds minimum_duration = std::chrono::milliseconds(100));
+
+  RangeTimerPtr createTimer(TimerCb callback);
+
+  void setScaleFactor(float scale_factor);
+
+protected:
+  class ScaledRangeTimer;
+
+  struct BucketedEnabledTimer {
+    BucketedEnabledTimer(ScaledRangeTimer& timer, MonotonicTime latest_trigger_time)
+        : timer(timer), latest_trigger_time(latest_trigger_time) {}
+    ScaledRangeTimer& timer;
+    MonotonicTime latest_trigger_time;
+  };
+
+  using BucketEnabledList = std::list<BucketedEnabledTimer>;
+
+  void enqueuePendingTimer(ScaledRangeTimer& timer);
+  void disablePendingTimer(ScaledRangeTimer& timer);
+
+  BucketEnabledList::iterator add(ScaledRangeTimer& timer, MonotonicTime::duration max_duration);
+  void disableActiveTimer(MonotonicTime::duration max_duration,
+                          const BucketEnabledList::iterator& bucket_position);
+
+private:
+  struct Bucket {
+    TimerPtr timer;
+    BucketEnabledList scaled_timers;
```

Comment:
Should you be using a priority queue here in order to get more predictable behavior?

Think about the behavior of timers that are on the lower end of the bucket range when the scale factor is 1.0

Reply:
I don't understand, what would a priority queue get us?

Reply:
Use of a priority queue would provide sorted scaled timer behavior similar to that provided by a sorted multiset but with lower constant overhead.  What I was going for is sorting the timers in the bucket, and use of the actual remaining max-min instead of use of max for all elements inserted into a bucket.